ts/test/session/sending/MessageQueue_test.ts

```diff
@@ -0,0 +1,202 @@
+  let messageQueueStub: MessageQueue;
+  // Message Sender Stubs
+  let sendStub: sinon.SinonStub<[RawMessage, (number | undefined)?]>;
+  let sendToOpenGroupStub: sinon.SinonStub<[OpenGroupMessage]>;
+  // Group Utils Stubs
+  let isMediumGroupStub: sinon.SinonStub<[PubKey], boolean>;
+  let groupMembersStub: sinon.SinonStub;
+  // Session Protocol Stubs
+  let hasSessionStub: sinon.SinonStub<[PubKey], Promise<boolean>>;
+
+  beforeEach(async () => {
+    // Stub out methods which touch the database
+    const storageID = 'pendingMessages';
+    data = {
+      id: storageID,
+      value: '[]',
+    };
+
+    // Pending Message Cache Data Stubs
+    TestUtils.stubData('getItemById')
+      .withArgs('pendingMessages')
+      .callsFake(async () => {
+        return data;
+      });
+    TestUtils.stubData('createOrUpdateItem').callsFake((item: StorageItem) => {
+      if (item.id === storageID) {
+        data = item;
+      }
+    });
+
+    // Utils Stubs
+    sandbox.stub(UserUtil, 'getCurrentDevicePubKey').resolves(ourNumber);
+    TestUtils.stubData('getPairedDevicesFor').callsFake(async () => {
+      return generateMemberList(2);
+    });
+    TestUtils.stubWindow('libsignal', {
+      SignalProtocolAddress: sandbox.stub(),
+      SessionCipher: Stubs.SessionCipherStub,
+    } as any);
+
+    // Message Sender Stubs
+    sendStub = sandbox.stub(MessageSender, 'send').resolves();
+    sendToOpenGroupStub = sandbox
+      .stub(MessageSender, 'sendToOpenGroup')
+      .resolves(true);
+
+    // Group Utils Stubs
+    isMediumGroupStub = sandbox
+      .stub(GroupUtils, 'isMediumGroup')
+      .resolves(false);
+    groupMembersStub = sandbox
+      .stub(GroupUtils, 'getGroupMembers' as any)
+      .callsFake(async () => generateMemberList(10));
+
+    // Session Protocol Stubs
+    hasSessionStub = sandbox.stub(SessionProtocol, 'hasSession').resolves(true);
+    sandbox.stub(SessionProtocol, 'sendSessionRequest').resolves();
+    sandbox
+      .stub(SessionProtocol, 'sendSessionRequestIfNeeded')
+      .callsFake(async (pubkey: PubKey) => {
+        pubkey;
+        sessionRequestSent = true;
+      });
+
+    // Pending Mesage Cache Stubs
+    const chatMessages = Array.from({ length: 10 }, generateChatMessage);
+    const rawMessage = toRawMessage(
+      generateFakePubkey(),
+      generateChatMessage()
+    );
+
+    sandbox.stub(PendingMessageCache.prototype, 'add').resolves(rawMessage);
+    sandbox.stub(PendingMessageCache.prototype, 'remove').resolves();
+    sandbox
+      .stub(PendingMessageCache.prototype, 'getDevices')
+      .returns(generateMemberList(10));
+    sandbox
+      .stub(PendingMessageCache.prototype, 'getForDevice')
+      .returns(chatMessages.map(m => toRawMessage(generateFakePubkey(), m)));
+
+    messageQueueStub = new MessageQueue();
+  });
+
+  afterEach(() => {
+    TestUtils.restoreStubs();
+    sandbox.restore();
+  });
+
+  it('can send to a single device', async () => {
```

Comment:
i would split these tests up into categories such as the function they're testing.
E.g
```
describe('send') {
    it(...)
}

describe('sendMessageToDevices') {
    it(...)
}
```
this way in the future if we run into a bug in the function we can add it to the relevant category without having to worry about duplicated tests